app/src/main/java/com/keylesspalace/tusky/fragment/NotificationsFragment.kt

```diff
@@ -0,0 +1,688 @@
+        val adapter =
+            ArrayAdapter(context, android.R.layout.simple_list_item_multiple_choice, list)
+        val window = PopupWindow(context)
+        val view = LayoutInflater.from(context)
+            .inflate(R.layout.notifications_filter, view as ViewGroup?, false)
+        val listView = view.findViewById<ListView>(R.id.listView)
+        view.findViewById<View>(R.id.buttonApply)
+            .setOnClickListener {
+                val checkedItems = listView.checkedItemPositions
+                val excludes: MutableSet<Notification.Type> = HashSet()
+                for (i in notificationsList.indices) {
+                    if (!checkedItems[i, false]) excludes.add(notificationsList[i])
+                }
+                window.dismiss()
+                if (viewModel.uiState.value.activeFilter != excludes) {
+                    viewModel.accept(InfallibleUiAction.ApplyFilter(excludes))
+                }
+            }
+        listView.adapter = adapter
+        listView.choiceMode = ListView.CHOICE_MODE_MULTIPLE
+        for (i in notificationsList.indices) {
+            if (!viewModel.uiState.value.activeFilter.contains(notificationsList[i])) {
+                listView.setItemChecked(i, true)
+            }
+        }
+        window.contentView = view
+        window.isFocusable = true
+        window.width = ViewGroup.LayoutParams.WRAP_CONTENT
+        window.height = ViewGroup.LayoutParams.WRAP_CONTENT
+        window.showAsDropDown(binding.buttonFilter)
+    }
+
+    private fun getNotificationText(type: Notification.Type): String {
+        return when (type) {
+            Notification.Type.MENTION -> getString(R.string.notification_mention_name)
+            Notification.Type.FAVOURITE -> getString(R.string.notification_favourite_name)
+            Notification.Type.REBLOG -> getString(R.string.notification_boost_name)
+            Notification.Type.FOLLOW -> getString(R.string.notification_follow_name)
+            Notification.Type.FOLLOW_REQUEST -> getString(R.string.notification_follow_request_name)
+            Notification.Type.POLL -> getString(R.string.notification_poll_name)
+            Notification.Type.STATUS -> getString(R.string.notification_subscription_name)
+            Notification.Type.SIGN_UP -> getString(R.string.notification_sign_up_name)
+            Notification.Type.UPDATE -> getString(R.string.notification_update_name)
+            Notification.Type.REPORT -> getString(R.string.notification_report_name)
+            else -> "Unknown"
+        }
+    }
+
+    override fun onViewTag(tag: String) {
+        super.viewTag(tag)
+    }
+
+    override fun onViewAccount(id: String) {
+        super.viewAccount(id)
+    }
+
+    override fun onMute(mute: Boolean, id: String, position: Int, notifications: Boolean) {
+        // No muting from notifications yet
+    }
+
+    override fun onBlock(block: Boolean, id: String, position: Int) {
+        // No blocking from notifications yet
```

Comment:
Does that actually work? Because this is direct from the Java code, https://github.com/tuskyapp/Tusky/blob/develop/app/src/main/java/com/keylesspalace/tusky/fragment/NotificationsFragment.java#L759

Reply:
I've made it call `adapter.refresh()`. Although it does depend on how fast the server is. E.g., muting someone and then refreshing notifications will sometimes still show the notification from them.